rmw_connext_cpp/src/rmw_publisher.cpp

```diff
@@ -441,80 +427,71 @@ rmw_return_loaned_message_from_publisher(
     publisher handle,
-    publisher->implementation_identifier, rti_connext_identifier,
-    return RMW_RET_ERROR)
+    publisher->implementation_identifier,
+    rti_connext_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
 
+  rmw_ret_t ret = RMW_RET_OK;
   auto node_info = static_cast<ConnextNodeInfo *>(node->data);
-  if (!node_info) {
-    RMW_SET_ERROR_MSG("node info handle is null");
-    return RMW_RET_ERROR;
-  }
   auto participant = static_cast<DDS::DomainParticipant *>(node_info->participant);
-  if (!participant) {
-    RMW_SET_ERROR_MSG("participant handle is null");
-    return RMW_RET_ERROR;
-  }
   // TODO(wjwwood): need to figure out when to unregister types with the participant.
   ConnextStaticPublisherInfo * publisher_info =
     static_cast<ConnextStaticPublisherInfo *>(publisher->data);
-  if (publisher_info) {
-    node_info->publisher_listener->remove_information(
-      publisher_info->dds_publisher_->get_instance_handle(), EntityType::Publisher);
-    node_info->publisher_listener->trigger_graph_guard_condition();
-    DDS::Publisher * dds_publisher = publisher_info->dds_publisher_;
-
-    if (dds_publisher) {
-      if (publisher_info->topic_writer_) {
-        if (dds_publisher->delete_datawriter(publisher_info->topic_writer_) != DDS::RETCODE_OK) {
-          RMW_SET_ERROR_MSG("failed to delete datawriter");
-          return RMW_RET_ERROR;
-        }
-        publisher_info->topic_writer_ = nullptr;
-      }
-      if (participant->delete_publisher(dds_publisher) != DDS::RETCODE_OK) {
-        RMW_SET_ERROR_MSG("failed to delete publisher");
-        return RMW_RET_ERROR;
-      }
-      publisher_info->dds_publisher_ = nullptr;
-    } else if (publisher_info->topic_writer_) {
-      RMW_SET_ERROR_MSG("cannot delete datawriter because the publisher is null");
-      return RMW_RET_ERROR;
+  node_info->publisher_listener->remove_information(
+    publisher_info->dds_publisher_->get_instance_handle(), EntityType::Publisher);
+  node_info->publisher_listener->trigger_graph_guard_condition();
+  DDS::Publisher * dds_publisher = publisher_info->dds_publisher_;
+
+  if (dds_publisher->delete_datawriter(publisher_info->topic_writer_) != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("failed to delete datawriter");
+    ret = RMW_RET_ERROR;
+  }
+  if (participant->delete_publisher(dds_publisher) != DDS::RETCODE_OK) {
+    if (!rmw_error_is_set()) {
+      RMW_SET_ERROR_MSG("failed to delete publisher");
+      ret = RMW_RET_ERROR;
+    } else {
+      RMW_SAFE_FWRITE_TO_STDERR("failed to delete publisher\n");
     }
+  }
 
-    ConnextPublisherListener * pub_listener = publisher_info->listener_;
-    if (pub_listener) {
-      RMW_TRY_DESTRUCTOR(
-        pub_listener->~ConnextPublisherListener(),
-        ConnextPublisherListener, return RMW_RET_ERROR)
-      rmw_free(pub_listener);
-    }
+  ConnextPublisherListener * pub_listener = publisher_info->listener_;
 
+  if (!rmw_error_is_set()) {
```

Comment:
Same as in other equivalent PRs, I don't like having to do this to ensure the first error that occurs is the one that pops up and that no subsequent error goes unnoticed.